File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        if(n==1) return nums[0];
        int[] first_half = new int[n];
        int[] last_half = new int[n];

        for(int i=0;i<n-1;i++){
            first_half[i] = nums[i];
            last_half[i] = nums[i+1];
        }

        return Math.max(robpart(first_half), robpart(last_half));
    }

    public int robpart(int[] nums){
        int n = nums.length;
        if(n==1) return nums[0];
        int[] dp = new int[nums.length+1];
        dp[0] = nums[0];
        dp[1] = Math.max(nums[0],nums[1]);
        for(int i=2;i<n;i++){
            dp[i] = Math.max(nums[i]+dp[i-2],dp[i-1]);
        }
        return dp[n-1];
    }
}

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n<=0:
            return 0
        if n==1:
            return nums[0] 
        def rob_linear(nums):
            dp = [0] * len(nums)
            if len(nums)==1:
                return nums[0] 
            dp[0] = nums[0]
            dp[1] = max(nums[0], nums[1])
            for i in range(2, len(nums)):
                dp[i] = max(dp[i-1], nums[i] + dp[i-2])
            return dp[len(nums)-1]
        return max(rob_linear(nums[:-1]), rob_linear(nums[1:]))

